distributed/actor.py

```diff
@@ -241,22 +243,32 @@ def __init__(self, q, io_loop, result=None):
         self.io_loop = io_loop
         if result:
             self._cached_result = result
+        self.status = "pending"
 
     def __await__(self):
         return self.result()
 
-    def result(self, timeout=None):
-        try:
-            if isinstance(self._cached_result, Exception):
-                raise self._cached_result
-            return self._cached_result
-        except AttributeError:
-            out = self.q.get(timeout=timeout)
+    async def _result(self, raiseit=True):
+        if not hasattr(self, "_cached_result"):
+            out = await self.q.get()
             if out["status"] == "OK":
                 self._cached_result = out["result"]
             else:
                 self._cached_result = out["exception"]
-        return self.result()
+        if isinstance(self._cached_result, Exception):
+            self.status = "error"
+            raise self._cached_result
```

Comment:
If the return value is an exception then this fails inappropriately.  We probably want to condition on `out["status"]` instead.